src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -3544,11 +3521,35 @@ class G1PrepareEvacuationTask : public AbstractGangTask {
 };
 
 void G1CollectedHeap::pre_evacuate_collection_set(G1EvacuationInfo* evacuation_info, G1ParScanThreadStateSet* per_thread_states) {
+  // Please see comment in g1CollectedHeap.hpp and
+  // G1CollectedHeap::ref_processing_init() to see how
+  // reference processing currently works in G1.
+  _ref_processor_stw->start_discovery(false /* always_clear */);
+
   _bytes_used_during_gc = 0;
 
   _expand_heap_after_alloc_failure = true;
   Atomic::store(&_num_regions_failed_evacuation, 0u);
 
+  wait_for_root_region_scanning();
```

Comment:
I kind of dislike having this wait moved here into this helper function, where I feel like it's less obvious in the flow of the collection pause.  I'd rather it was still in the main body.  And it's not like there's anything gained by waiting until after the other things that precede it in this function.  It looks like it's okay to delay the wait later than its current location though; the scan is performed using the concurrent marking threads, so other parts of the setup along the way should be fine.

Reply:
I came to the same conclusion and considering that waiting as a candidate for putting it into the going-to-be parallel pre-evacuation phase. Then again, it might not be useful to have both the concurrent worker threads and parallel threads working at the same time (for performance reasons only; I do not consider the potential parallel verification also using the parallel worker threads as an issue). I checked that verification does not interfere with it (i.e. verification does not care about the mark bitmap in that sense).
The only requirement I can see is that all of these root-region objects need to have their referents marked, i.e. they must not be overwritten in some way, which this place guarantees as well, and to me it seems a more fitting place.

Reply:
A problem with moving it later to the new placement is that it's now included in the pause time, where it previously wasn't.  I'm not sure it should be included in the pause time, as it's not really part of this collection.  I also wonder if the root scanning ought to be done using the concurrent workers rather than the stw workers, as there are typically fewer of the former (required to not be more?).

Reply:
It is true that it is now also included in the time used for MMU calculation (as part of the `record_collection_start/end` pair), which may a problem.

Previous gc timing also included it though, i.e. in the `note_gc_start()/print_phase_times()` bracket.

I'll move it back then.

Reply:
Things are unfortunately a bit messy and confusing/confused about pause time handling; see JDK-8240779 and JDK-7178365.

Reply:
I am aware of those, but this change does not really intend to fix these issues, but hopefully not make them worse (afaict this has been achieved).